File: LeetCode Monthly/Feb (2023)/783.ts
//https://leetcode.com/problems/minimum-distance-between-bst-nodes/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function minDiff(root: TreeNode | null): number {
  // With List
  return minDiffInBST1(root)
  // No List
  return minDiffInBST2(root)
}

function minDiffInBST1(root: TreeNode | null): number {
  let minDiff: number = Infinity
  let list: number[] = []

  orderList(root)
  function orderList(node: TreeNode | null) {
    if (!node) return
    orderList(node.left)
    list.push(node.val)
    orderList(node.right)
  }

  for (let i = 1; i < list.length; i++) {
    minDiff = Math.min(minDiff, list[i] - list[i - 1])
  }

  return minDiff
}

function minDiffInBST2(root: TreeNode | null): number {
  let minDiff: number = Infinity
  let preValue: TreeNode = null

  function orderList(node: TreeNode | null) {
    if (node === null) return

    orderList(node.left)

    if (preValue !== null) {
      minDiff = Math.min(minDiff, node.val - preValue.val)
    }
    preValue = node

    orderList(node.right)
  }

  orderList(root)
  return minDiff
}

File: LeetCode Monthly/Feb (2023)/783.py
# https://leetcode.com/problems/minimum-distance-between-bst-nodes/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        minDiff: int = float('inf')
        list = []

        def orderList(node):
            if not node: return

            orderList(node.left)
            list.append(node.val)
            orderList(node.right)

        orderList(root)

        for i in range(len(list) - 1):
            minDiff = min(minDiff, list[i + 1] - list[i])

        return minDiff
        
